tests/func/test_ignore.py

```diff
@@ -56,3 +55,76 @@ def test_ignore_in_parent_dir(self):
         all_paths = self._get_all_paths()
 
         self.assertNotIn(forbidden_path, all_paths)
+
+
+def test_metadata_unchanged_when_moving_ignored_file(dvc_repo, repo_dir):
+    new_data_path = repo_dir.DATA_SUB + "_new"
+
+    ignore_file = os.path.join(dvc_repo.root_dir, DvcIgnore.DVCIGNORE_FILE)
+    repo_dir.create(
+        ignore_file,
+        "\n".join(
+            [to_posixpath(repo_dir.DATA_SUB), to_posixpath(new_data_path)]
+        ),
+    )
+
+    mtime_sig, size = get_mtime_and_size(repo_dir.DATA_DIR)
+
+    shutil.move(repo_dir.DATA_SUB, new_data_path)
+
+    new_mtime_sig, new_size = get_mtime_and_size(repo_dir.DATA_DIR)
+
+    assert new_mtime_sig == mtime_sig
+    assert new_size == size
+
+
+def test_mtime_changed_when_moving_non_ignored_file(dvc_repo, repo_dir):
+    new_data_path = repo_dir.DATA_SUB + "_new"
+    mtime, size = get_mtime_and_size(repo_dir.DATA_DIR)
+
+    shutil.move(repo_dir.DATA_SUB, new_data_path)
+    new_mtime, new_size = get_mtime_and_size(repo_dir.DATA_DIR)
+
+    assert new_mtime != mtime
+    assert new_size == size
+
+
+def test_metadata_unchanged_on_ignored_file_deletion(dvc_repo, repo_dir):
+    ignore_file = os.path.join(dvc_repo.root_dir, DvcIgnore.DVCIGNORE_FILE)
+    repo_dir.create(ignore_file, to_posixpath(repo_dir.DATA_SUB))
+    mtime_sig, size = get_mtime_and_size(repo_dir.DATA_DIR)
+
+    os.remove(repo_dir.DATA_SUB)
+    new_mtime_sig, new_size = get_mtime_and_size(repo_dir.DATA_DIR)
+
+    assert new_mtime_sig == mtime_sig
+    assert new_size == size
+
+
+def test_metadata_changed_on_non_ignored_file_deletion(dvc_repo, repo_dir):
+    mtime_sig, size = get_mtime_and_size(repo_dir.DATA_DIR)
+
+    os.remove(repo_dir.DATA_SUB)
+    new_mtime_sig, new_size = get_mtime_and_size(repo_dir.DATA_DIR)
+
+    assert new_mtime_sig != mtime_sig
+    assert new_size != size
+
+
+def test_ignore_should_work_for_external_dependency(dvc_repo, repo_dir):
```

Comment:
Hm, this is an interesting one. So dvcignores work in repo for everything in repo and also for external dependencies and, i suppose, external outputs? So one would be able to add .dvcignore inside of his output directory and dvc would use it? Also, do things like `*.pyc`(so any *.pyc file, with no specific path, don't confuse it with `/*.pyc`) and stuff in the .dvcignore in the root of the repo work for external dependencies and outputs?

Reply:
As I see repos root `.dvcignore` is not used when outside of repo.

Reply:
@Suor is right.
I see one problem with such use case: if we start to support it, we would probably have to be prepare to handle ignores in particular project from `.dvcignore` laying anywhere inside the project, thus switching from loading `.dvcignore` during traversing, to loading all dvcignores before traversing any directory.

Other way would be probably to restrict where we can put ignores for external dirs. (eg. root of the project)

Reply:
i'm not asking to support it, I'm just asking if it is the intension or not :slightly_smiling_face: But the `.dvcignore` in output directories works, right?

Reply:
If `.dvcignore` inside of an output directory is supported, then consider this scenario:

1) `dvc run` outputs a new directory and places .dvcignore in it; We use `.dvcignore` to not cache what is says and it itself.
2) now we clean everything up and run `dvc checkout`, which will not create `dir/.dvcignore` and so if anything creates some ignored file in there, it no longer will be ignored. E.g. imagine `*.swp` files created, when you decide to open one of the output files with vim. Is this an intended behavior? Should we not even use `.dvcignore` that is located inside of an output dir? or do we ignore it already?

Reply:
That is good question, `.dvcignore` is ignored by default. In scenario described by you we will remove it and therefore make "change" in data directory. That is defienitely not desired behaviour. 
I see here 3 ways how we can proceed:

- we should store `.dvcignore` with DVC (and not scm) which will require user to commit data directory when creating `.dvcignore` file inside. Semantically that seems ok to me, as `.dvcignore` seems more like a part of dvc workflow and not "git". On the other hand that raises some problems, like how to add it? `.dvcignore` will be tightly binded to particular step of pipeline (like adding dataset that contains unwanted parts). That would mean we would probably need to add it with particular step of creating ML pipeline, which would probably require new option like `dvc add data --ignore /path/to/project/.dvcignore`. That would araise new problems (like what if 2 steps require same `.dvcignore` file)

-  we can restrict that ignores for external outputs can be written only in some particular place (eg workdir root), that behaviour will spawn new problems (eg how to write ignore for remotes different than local, I guess we would need to do it using url) 

- maybe we should forbid external outputs `.dvcignore`'s

Reply:
@pared But the problem that I've described is not only related to external outputs but to in-repo ones as well, right? E.g. you have `dvc run -o data_dir` which has `data_dir/.dvcignore`. Is `data_dir/.dvcignore` going to be parsed by dvc?

Reply:
> EDIT:
> after giving it some thought, I don't think that described problem exists. Why?
> ...

but your case doesn't involve adding previously ignored files (e.g. `*.swp`) to the directory later. That means that `dvc` will no longer ignore them because `.dvcignore` file is gone now. My point is that in order to avoid that confusion when at the begging some files are ignored and then they are not, we should simply not parse `.dvcignore`s that are anywhere inside of the output directory. This is similar to how we treat *.dvc files in output directories - we simply ignore them when collecting the DAG.

Reply:
@efiop I agree with your point, we should forbid placing `.dvcignore` inside outs.
In that case, if we want to support `.dvcignore` for external dependencies, it would have to be handled from the repository. I think that we should create new issue for this problem, and discuss it there. what do you think?

Reply:
@pared We can just skip or even raise an error if we find `.dvcignore` there. I guess with the current way of handling everything, both approaches would require the same amount of work, right? If we would go with the `raise and arror` route, we could maybe also raise an error there when we find `.dvc` files :slightly_smiling_face: I think one of the approaches should be adopted within this PR, so that we don't create the confusion discussed above.

Reply:
As to supporting .dvcignore for external outputs/deps in general, I see a few appraoches:

1) Use rules defined in the root of the repo for external deps and outs. This kinda makes sense if you think about external outs/deps as a continuation of your repo workspace. But at the same time, if rules like `*.swp` are quite straightforward, it surely creates a confusion when you want to ignore something by an abs path `/path/to/*.swp`, since in git terms it would actually meen `repo_root/path/to/*.swp`;

2) Support system-level and globa-level dvcignores, like git does. This one is actually pretty straightforward and makes totall sense for local files(e.g. not s3 or ssh).

3) Invent some syntax for the dvcignore to handle external deps/outs or even a separate file like `.dvcignore-external` or something like that. This one is tricky, but at least it could also be applied to any s3/ssh/etc deps/outs, since we would have to treat those separately anyway.

But yeah, these are just some thoughts, I would not bother with it until we see the demand.

Reply:
What is the effort of just ignoring `.dvcignore` files?

@efiop system/global dvcignores is a no go since they will be an invisible dependencies for all stages - when you change one `dvc repro` will result in different thing than `dvc repro -f`.

Reply:
great point about dependencies @Suor !

Reply:
@Suor Although, in that case, our local `.dvcignore`s are also dependencies :) We shouldn't treat .dvcignores as dependencies at all, because the scenarios that they are used in could totally wait for the next repro to ignore listed files. I guess this is the case where we could sacrifice dep logic. Or maybe we should actually re-evaluate dvcignores and make them distributes across dvc-files,? The funny thing is that this problem is the same that we have with locks for imported external repos :slightly_smiling_face:

Reply:
Though, unlike locks, dvcignores are clearly a global concept and I can totally see people requesting the current global dvcignore later and I don't see them requesting per-dvcfile dvcignore. So I would stick with the current idea of dvcignore.